File: Java/Bubblesort.java
// Introduction
------------------
// Bubble Sort is the simplest sorting algorithm that works by repeatedly
// swapping the adjacent elements if they are in wrong order.
------------------
// TIME COMPLEXITIES
------------------------
// Worst-case performance	O(n^2)
// Best-case performance	O(n)
// Average performance	    O(n^2)
------------------------

// SPACE COMPLEXITIES
------------------------
// Worst-case performance	O(1)
------------------------


package Sorts;

import static Sorts.SortUtils.*;


class BubbleSort implements SortAlgorithm {

  @Override
  public <T extends Comparable<T>> T[] sort(T[] array) {
    for (int i = 0, size = array.length; i < size - 1; ++i) {
      boolean swapped = false;
      for (int j = 0; j < size - 1 - i; ++j) {
        if (greater(array[j], array[j + 1])) {
          swap(array, j, j + 1);
          swapped = true;
        }
      }
      if (!swapped) {
        break;
      }
    }
    return array;
  }

  /** Driver Code */
  public static void main(String[] args) {

    Integer[] integers = {4, 23, 6, 78, 1, 54, 231, 9, 12};
    BubbleSort bubbleSort = new BubbleSort();
    bubbleSort.sort(integers);

    for (int i = 0; i < integers.length - 1; ++i) {
      assert integers[i] <= integers[i + 1];
    }
    print(integers); /* output: [1, 4, 6, 9, 12, 23, 54, 78, 231] */

    String[] strings = {"c", "a", "e", "b", "d"};
    bubbleSort.sort(strings);
    for (int i = 0; i < strings.length - 1; i++) {
      assert strings[i].compareTo(strings[i + 1]) <= 0;
    }
    print(bubbleSort.sort(strings)); /* output: [a, b, c, d, e] */
  }
}

File: Golang/BubbleSort.go
package main

import "fmt"

/*

BUBBLE SORT ALGORITHM
--------------------------
Time complexity
-Best case  O(n)
-worst case O(n^2)
---------------------------
Space complexity - O(1)

*/

/*
Function explanation:
We simply iterate through array and repedtedly swap the elements if they are in wrong order
*/
func BubbleSort(arr []int, n int) {
	sorted := false
	for !sorted {
		swapped := false
		for i := 0; i < n-1; i++ {
			if arr[i] > arr[i+1] {
				arr[i+1], arr[i] = arr[i], arr[i+1]
				swapped = true
			}
		}
		if !swapped {
			sorted = true
		}
		n--
	}
}

func main() {
	var size int
	fmt.Scanln(&size)
	input := make([]int, size)
	for i := 0; i < size; i++ {
		fmt.Scan(&input[i])
	}
	BubbleSort(input, size)
	fmt.Println(input)
}
